lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,311 @@
+        start += (end - start) / 2
+
+        # Find the first activation
+        start = df[(df.index >= start) &
+                   (df['active'] == 1)].first_valid_index()
+
+        if (start >= end):
+            raise ValueError('Phase ends before it has even started')
+
+        df = df.ffill()
+        df = df_window(df, (start, end), method='exclusive')
+
+        # Fake Time, representing only the task activation period
+        df_add_delta(df, col='Time', inplace=True)
+        df = df.dropna()
+        df = df[df.active == 1]
+        df['Time'] = df['Time'].cumsum()
+
+        # Frequency to capacity
+        capacities = self._collect_capacities(self.plat_info)
+        for cpu, freq_to_capa in capacities.items():
+            df[cpu] = df[cpu].map(freq_to_capa)
+
+        df['capacity'] = df.apply(lambda line: line[line.cpu], axis=1)
+        df['capacity_max'] = df.apply(
+                                lambda line:
+                                max(capacities[line.cpu].values()),
+                                axis=1
+                            )
+
+        df.set_index('Time', inplace=True)
+
+        return (
+                series_mean(df['capacity']),
+                df['capacity_max'].max(),
+                df.cpu.unique()[-1]
+        )
+
+    @memoized
+    def _get_test_df(self):
+        task = self.rtapp_task_ids_map['task'][0]
+
+        df = self.trace.analysis.rta.df_phases(task)
+        df = df.reset_index()
+        df.rename(columns={'index': 'start'}, inplace=True)
+        df['end'] = df['start'].shift(-1)
+
+        # TODO: To remove once we have named phases.
+        df = df[df.phase > 0]
+
+        df[['capacity', 'capacity_max', 'cpu']] = df.apply(
+                lambda x:
+                self._get_task_activation_mean_capacities(
+                    task, x['start'], x['end']),
+                axis=1, result_type='expand')
+
+        phases = [(0, 0, 0)] + self._get_phases(self.plat_info)
+
+        df[['uclamp_val', 'util', 'band']] = df.apply(
+                lambda x: [phases[int(x.phase)][0],
+                           phases[int(x.phase)][1],
+                           phases[int(x.phase)][2]],
+                axis=1, result_type='expand')
+
+        return df
+
+    @classmethod
+    def _test_passed(cls, df, col, minmax_func):
+        minmax_df = df.apply(minmax_func, axis=1, result_type='expand')
+        passed = df[col].between(minmax_df[0], minmax_df[1])
+
+        return ResultBundle.from_bool(passed.all())
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_placement(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement is compatible with
+        UtilClamp requirements. This is done by comparing the maximum capacity
+        of the CPU on which the task has been placed, with the UtilClamp
+        value.
+        """
+
+        df = self._get_test_df()
```

Comment:
I'm not fond of what `_get_task_activation_mean_capacities()` does under the hood. We don't need to compute any of the capacity means here, do we? Couldn't we simplify this test to something like:

```
task = self.rtapp_task_ids_map['task'][0]
df_activations = self.trace.analysis.tasks.df_task_activation(task)

for start, end, uclamp_val in XXX:
    df = df_activations[start:end]['cpu']
    fitting_cpus = [cpu for cpu, cap in self.plat_info['cpu-capacities']['orig'].items() if cap > uclamp_val * 1.25]
    df_notfit = df[~df['cpu'].isin(fitting_cpus)]
    if df_notfit.any():
        # add metrics, store e.g. first timestamp failure, maybe event plot them

```

Reply:
Couldn't reuse all the proposal here, because I need some common function that slice the phases. However I reused the fitting_cpus list idea. With a set, that's even simpler than it used to be, to check if actual CPUs are a subset of the expected and if not, what are the wrong ones.